lib/app.d.ts

```diff
@@ -0,0 +1,147 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+declare namespace Realm {
+    /**
+     * The constructor of MongoDB Realm App.
+     */
+    type AppConstructor = new <FF extends FunctionsFactory>(
+        id: string,
+        configuration?: Partial<AppConfiguration>,
+    ) => App<FF>;
+
+    /**
+     * A MongoDB Realm App.
+     */
+    interface App<FF extends FunctionsFactory = FunctionsFactory> {
+        /**
+         * The id of this Realm app.
+         */
+        id: string;
+
+        /**
+         * Use this to call functions defined on the MongoDB Realm server.
+         */
+        functions: FF;
+
+        /**
+         * The last user to log in or being switched to.
+         */
+        currentUser: Realm.User | null;
+
+        /**
+         * All authenticated users.
+         */
+        allUsers: Readonly<Realm.User[]>;
+
+        /**
+         * Log in a user using a specific credential
+         * @param credentials the credentials to use when logging in
+         */
+        logIn(credentials: Credentials): Promise<Realm.User>;
+
+        /**
+         * Log out the currently authenticated user and clear any persisted authentication information.
+         */
+        logOut(): Promise<void>;
+
+        /**
+         * Switch current user, from an instance of `Realm.User` or the string id of the user.
+         */
+        switchUser(user: User | string): void;
+    }
+
+    interface AppConfiguration {
+        baseUrl: string;
+    }
+
+    interface Credentials {
+        /**
+         * Name of the authentication provider.
+         */
+        readonly providerName: string;
+    
+        /**
+         * Type of the authentication provider.
+         */
+        readonly providerType: string;
+    
+        /**
+         * The contents of this credential as they will be passed to the server.
+         */
+        readonly material: { [key: string]: string };
```

Comment:
I agree - I just took this name from the existing Stitch SDK (didn't know if it would carry over to something in a Stitch request or something). But .. it turns out its a name which is totally internal to the SDK, I'll rename it.

Reply:
Actually - I believe this is the intent of the `toJSON` method called by `JSON.stringify`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON_behavior: I'll change it into a method instead.

Reply:
👍